numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -125,27 +196,34 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
     if (len < 0) {
         PyErr_SetString(PyExc_TypeError,
                 "expected sequence object with len >= 0 or a single integer");
+        Py_DECREF(seq_obj);
```

Comment:
`PySequence_Fast_GET_SIZE` will never fail. You can remove the whole `< 0` and `== -1` blocks.  Any failure will have occurred in `PySequence_Fast` already.

We may have to think briefly whether we have to fall back to the above branch if  `PySequence_Fast(...)` fails, so that it is a bit like a: `try: list(obj); except: int(obj)` path in Python.

Reply:
The reason why this line is not covered, is that is is impossible to cover it :).

Reply:
I removed the redundant checks in e5b77c1.

> We may have to think briefly whether we have to fall back to the above branch if PySequence_Fast(...) fails

Which are the conditions under which `PySequence_Fast` might fail in this case?

> The reason why this line is not covered, is that is is impossible to cover it :).

Got it :)

Reply:
`PySequence_Fast` should fail exactly in the same cases as `list(obj)` failing.

Reply:
Isn't this check sufficient to ensure that `PySequence_Fast` does not fail?
https://github.com/numpy/numpy/blob/5cbde7fef71268fc0b834c5f82aee798c6cb0792/numpy/core/src/multiarray/conversion_utils.c#L152

Reply:
No, for one, `PySequence_Check` only checks very quickly if it should be a squence.  But sequences could be badly implemented and only error when you get the elements.

The one problem that might occur is to try the sequence conversion first. Then we will still try the `int(seq)` path for broken sequences.  This is probably actually important!  Could you add a test for `np.array(3)` being accepted?  
```
seq = NULL
if not Long_CheckExact and PySequence_Check:
   seq = ...
   if seq == NULL:
       PyErr_Clear()  # ignore the error.

if seq == NULL:
    # integer path

else:
   # sequence path
```
A bit annoying dance, but I think it ensures we don't accidentally change more behaviour than anticipated. The `PyErr_Clear()` makes it a `try/except` clause.

Reply:
Please @seberg have a look at 6da32a8. 

For some reason now the following test fails (everything else is fine):
```
    def test_none(self):
        # once the warning expires, this will raise TypeError
        with pytest.warns(DeprecationWarning):
>           assert self.conv(None) == ()
E           assert None == ()
E            +  where None = <built-in function run_intp_converter>(None)
E            +    where <built-in function run_intp_converter> = <numpy.core.tests.test_conversion_utils.TestIntpConverter object at 0x7fda6307ecd0>.conv

self       = <numpy.core.tests.test_conversion_utils.TestIntpConverter object at 0x7fda6307ecd0>

numpy/core/tests/test_conversion_utils.py:190: AssertionError
```

From what I understand, this case should be handled in:
https://github.com/numpy/numpy/blob/6da32a8349eba9f1dde56c12bdc347eda4ded5c7/numpy/core/src/multiarray/conversion_utils.c#L120-L128
But it looks like something went wrong.

Reply:
@seberg The failing test has been fixed in 83d8d59.